File: frontend/src/api/mcp.ts
import { apiClient } from './client'

// --- Types ---

export interface McpServerStatus {
  name: string
  healthy: boolean
  degraded: boolean
  transport: string | null
  tools_count: number
  avg_response_time_ms: number
  success_rate: number
  last_error: string | null
  last_check: string | null
  consecutive_failures: number
}

export interface McpStatusResponse {
  overall_status: string
  total_servers: number
  healthy_count: number
  unhealthy_count: number
  degraded_count: number
  total_tools: number
  servers: McpServerStatus[]
}

export interface McpToolInfo {
  name: string
  description: string | null
  server_name: string
  parameters: Record<string, unknown> | null
}

export interface McpToolListResponse {
  server_name: string | null
  tools: McpToolInfo[]
  total: number
}

export interface McpTestConnectionRequest {
  name: string
  transport: string
  command?: string | null
  args?: string[] | null
  url?: string | null
  env?: Record<string, string> | null
  headers?: Record<string, string> | null
}

export interface McpTestConnectionResponse {
  success: boolean
  latency_ms: number
  tools_count: number
  error: string | null
}

// --- SSE Event Data ---

export interface McpHealthEventData {
  event_id: string | null
  timestamp: number
  server_name: string
  healthy: boolean
  degraded: boolean
  error: string | null
  tools_available: number
  avg_response_time_ms: number
  success_rate: number
  last_check: string | null
}

// --- API Functions ---

export async function getMcpStatus(): Promise<McpStatusResponse> {
  const response = await apiClient.get<{ data: McpStatusResponse }>('/mcp/status')
  return response.data.data
}

export async function getMcpServers(): Promise<McpServerStatus[]> {
  const response = await apiClient.get<{ data: McpServerStatus[] }>('/mcp/servers')
  return response.data.data
}

export async function getMcpServerTools(
  serverName: string,
): Promise<McpToolListResponse> {
  const response = await apiClient.get<{ data: McpToolListResponse }>(
    `/mcp/servers/${encodeURIComponent(serverName)}/tools`,
  )
  return response.data.data
}

export async function testMcpConnection(
  config: McpTestConnectionRequest,
): Promise<McpTestConnectionResponse> {
  const response = await apiClient.post<{ data: McpTestConnectionResponse }>(
    '/mcp/test-connection',
    config,
  )
  return response.data.data
}

File: backend/app/interfaces/schemas/mcp.py
"""Pydantic response schemas for MCP REST endpoints."""

from pydantic import BaseModel, Field


class McpServerStatus(BaseModel):
    """Health status for a single MCP server."""

    name: str
    healthy: bool
    degraded: bool = False
    transport: str | None = None  # "stdio", "sse", "streamable_http"
    tools_count: int = 0
    avg_response_time_ms: float = 0.0
    success_rate: float = 100.0
    last_error: str | None = None
    last_check: str | None = None  # ISO timestamp
    consecutive_failures: int = 0


class McpStatusResponse(BaseModel):
    """Aggregated MCP health summary."""

    overall_status: str = "unknown"  # "healthy", "degraded", "unhealthy", "unknown"
    total_servers: int = 0
    healthy_count: int = 0
    unhealthy_count: int = 0
    degraded_count: int = 0
    total_tools: int = 0
    servers: list[McpServerStatus] = Field(default_factory=list)


class McpToolInfo(BaseModel):
    """Information about a single MCP tool."""

    name: str
    description: str | None = None
    server_name: str
    parameters: dict | None = None  # JSON Schema


class McpToolListResponse(BaseModel):
    """List of tools for one or all MCP servers."""

    server_name: str | None = None
    tools: list[McpToolInfo] = Field(default_factory=list)
    total: int = 0


class McpTestConnectionRequest(BaseModel):
    """Request body for test-connection endpoint."""

    name: str
    transport: str  # "stdio", "sse", "streamable_http"
    command: str | None = None  # for stdio
    args: list[str] | None = None  # for stdio
    url: str | None = None  # for sse / streamable_http
    env: dict[str, str] | None = None
    headers: dict[str, str] | None = None


class McpTestConnectionResponse(BaseModel):
    """Result of a test-connection attempt."""

    success: bool
    latency_ms: float = 0.0
    tools_count: int = 0
    error: str | None = None
